File: apps/memorai/src/sdk/memorai-sdk.ts
/**
 * MemorAI TypeScript SDK
 * Official client library for the MemorAI REST API
 * 
 * @version 1.0.0
 * @author MemorAI Team
 */

export interface MemorAIConfig {
  apiKey: string;
  baseURL?: string;
  timeout?: number;
  retries?: number;
  userAgent?: string;
}

export interface Memory {
  id?: string;
  structuredKey?: string;
  content: string;
  agentId: string;
  importance?: number;
  project?: string;
  tags?: string[];
  createdAt?: string;
  updatedAt?: string;
  metadata?: Record<string, any>;
}

export interface SearchOptions {
  query: string;
  limit?: number;
  project?: string;
  tags?: string[];
  importance?: { min?: number; max?: number };
  dateRange?: { from?: string; to?: string };
}

export interface ExportOptions {
  format?: 'json' | 'csv';
  project?: string;
  tags?: string[];
  dateFrom?: string;
  dateTo?: string;
}

export interface ImportOptions {
  skipDuplicates?: boolean;
  updateExisting?: boolean;
  preserveTimestamps?: boolean;
}

export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: {
    code: string;
    message: string;
    details?: any;
  };
  meta?: {
    timestamp: string;
    count?: number;
    cached?: boolean;
    [key: string]: any;
  };
}

export class MemorAIError extends Error {
  public code: string;
  public details?: any;

  constructor(message: string, code: string, details?: any) {
    super(message);
    this.name = 'MemorAIError';
    this.code = code;
    this.details = details;
  }
}

export class MemorAISDK {
  private config: Required<MemorAIConfig>;

  constructor(config: MemorAIConfig) {
    this.config = {
      baseURL: 'https://api.memorai.ro',
      timeout: 30000,
      retries: 3,
      userAgent: 'MemorAI-JS-SDK/1.0.0',
      ...config
    };

    if (!this.config.apiKey) {
      throw new MemorAIError('API key is required', 'MISSING_API_KEY');
    }
  }

  /**
   * Make HTTP request to MemorAI API
   */
  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<ApiResponse<T>> {
    const url = `${this.config.baseURL}${endpoint}`;
    const headers = {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${this.config.apiKey}`,
      'User-Agent': this.config.userAgent,
      ...options.headers
    };

    const requestOptions: RequestInit = {
      ...options,
      headers,
      signal: AbortSignal.timeout(this.config.timeout)
    };

    let lastError: Error | null = null;

    for (let attempt = 0; attempt < this.config.retries; attempt++) {
      try {
        const response = await fetch(url, requestOptions);

        if (!response.ok) {
          const errorData = await response.json().catch(() => ({}));
          throw new MemorAIError(
            errorData.error?.message || `HTTP ${response.status}: ${response.statusText}`,
            errorData.error?.code || 'HTTP_ERROR',
            { status: response.status, ...errorData.error?.details }
          );
        }

        const data = await response.json();
        return data as ApiResponse<T>;

      } catch (error) {
        lastError = error as Error;

        if (error instanceof MemorAIError) {
          throw error;
        }

        // Retry on network errors
        if (attempt === this.config.retries - 1) {
          throw new MemorAIError(
            `Request failed after ${this.config.retries} attempts: ${lastError.message}`,
            'NETWORK_ERROR',
            { originalError: lastError }
          );
        }

        // Wait before retry
        await new Promise(resolve => setTimeout(resolve, Math.pow(2, attempt) * 1000));
      }
    }

    throw lastError!;
  }

  /**
   * Create a new memory
   */
  async createMemory(memory: Omit<Memory, 'id' | 'createdAt' | 'updatedAt'>): Promise<Memory> {
    const response = await this.request<Memory>('/api/memories', {
      method: 'POST',
      body: JSON.stringify(memory)
    });

    if (!response.success || !response.data) {
      throw new MemorAIError(
        response.error?.message || 'Failed to create memory',
        response.error?.code || 'CREATE_FAILED'
      );
    }

    return response.data;
  }

  /**
   * Get all memories with optional filtering
   */
  async getMemories(options?: {
    category?: string;
    tags?: string[];
    limit?: number;
  }): Promise<Memory[]> {
    const params = new URLSearchParams();
    if (options?.category) params.set('category', options.category);
    if (options?.tags?.length) params.set('tags', options.tags.join(','));
    if (options?.limit) params.set('limit', options.limit.toString());

    const endpoint = `/api/memories${params.toString() ? `?${params}` : ''}`;
    const response = await this.request<Memory[]>(endpoint);

    if (!response.success || !response.data) {
      throw new MemorAIError(
        response.error?.message || 'Failed to fetch memories',
        response.error?.code || 'FETCH_FAILED'
      );
    }

    return response.data;
  }

  /**
   * Get a specific memory by ID
   */
  async getMemory(id: string): Promise<Memory> {
    const response = await this.request<Memory>(`/api/memories/${id}`);

    if (!response.success || !response.data) {
      throw new MemorAIError(
        response.error?.message || 'Failed to fetch memory',
        response.error?.code || 'FETCH_FAILED'
      );
    }

    return response.data;
  }

  /**
   * Update a memory
   */
  async updateMemory(id: string, updates: Partial<Memory>): Promise<Memory> {
    const response = await this.request<Memory>(`/api/memories/${id}`, {
      method: 'PUT',
      body: JSON.stringify(updates)
    });

    if (!response.success || !response.data) {
      throw new MemorAIError(
        response.error?.message || 'Failed to update memory',
        response.error?.code || 'UPDATE_FAILED'
      );
    }

    return response.data;
  }

  /**
   * Delete a memory
   */
  async deleteMemory(id: string): Promise<boolean> {
    const response = await this.request<{ deleted: boolean }>(`/api/memories/${id}`, {
      method: 'DELETE'
    });

    if (!response.success || !response.data) {
      throw new MemorAIError(
        response.error?.message || 'Failed to delete memory',
        response.error?.code || 'DELETE_FAILED'
      );
    }

    return response.data.deleted;
  }

  /**
   * Search memories
   */
  async searchMemories(options: SearchOptions): Promise<Memory[]> {
    const params = new URLSearchParams();
    params.set('q', options.query);
    if (options.limit) params.set('limit', options.limit.toString());
    if (options.project) params.set('project', options.project);
    if (options.tags?.length) params.set('tags', options.tags.join(','));
    if (options.importance?.min) params.set('importanceMin', options.importance.min.toString());
    if (options.importance?.max) params.set('importanceMax', options.importance.max.toString());
    if (options.dateRange?.from) params.set('dateFrom', options.dateRange.from);
    if (options.dateRange?.to) params.set('dateTo', options.dateRange.to);

    const response = await this.request<Memory[]>(`/api/search?${params}`);

    if (!response.success || !response.data) {
      throw new MemorAIError(
        response.error?.message || 'Failed to search memories',
        response.error?.code || 'SEARCH_FAILED'
      );
    }

    return response.data;
  }

  /**
   * Export memories
   */
  async exportMemories(options: ExportOptions = {}): Promise<Blob> {
    const params = new URLSearchParams();
    if (options.format) params.set('format', options.format);
    if (options.project) params.set('project', options.project);
    if (options.tags?.length) params.set('tags', options.tags.join(','));
    if (options.dateFrom) params.set('dateFrom', options.dateFrom);
    if (options.dateTo) params.set('dateTo', options.dateTo);

    const url = `${this.config.baseURL}/api/memories/export?${params}`;
    const headers = {
      'Authorization': `Bearer ${this.config.apiKey}`,
      'User-Agent': this.config.userAgent
    };

    const response = await fetch(url, { headers });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new MemorAIError(
        errorData.error?.message || `Export failed: ${response.statusText}`,
        errorData.error?.code || 'EXPORT_FAILED'
      );
    }

    return response.blob();
  }

  /**
   * Import memories
   */
  async importMemories(
    memories: Memory[],
    options: ImportOptions = {}
  ): Promise<{
    summary: {
      totalProcessed: number;
      imported: number;
      updated: number;
      skipped: number;
      errors: number;
    };
    details?: { errors: any[] };
  }> {
    const response = await this.request<any>('/api/memories/import', {
      method: 'POST',
      body: JSON.stringify({ memories, options })
    });

    if (!response.success || !response.data) {
      throw new MemorAIError(
        response.error?.message || 'Failed to import memories',
        response.error?.code || 'IMPORT_FAILED'
      );
    }

    return response.data;
  }

  /**
   * Get API health status
   */
  async getHealth(): Promise<{
    status: string;
    timestamp: string;
    version?: string;
  }> {
    const response = await this.request<any>('/api/health');

    if (!response.success || !response.data) {
      throw new MemorAIError(
        response.error?.message || 'Failed to get health status',
        response.error?.code || 'HEALTH_CHECK_FAILED'
      );
    }

    return response.data;
  }
}

// Default export
export default MemorAISDK;

// Named exports
export { MemorAISDK as MemorAI };

File: apps/memorai/src/sdk/memorai_sdk.py
"""
MemorAI Python SDK
Official Python client library for the MemorAI REST API

Usage:
    from memorai_sdk import MemorAI, Memory
    
    client = MemorAI(
        api_key="your-api-key",
        base_url="https://api.memorai.ro"
    )
    
    # Create a memory
    memory = client.create_memory(
        content="Your memory content",
        agent_id="your-agent-id",
        tags=["ai", "memory"]
    )
    
    # Search memories
    results = client.search_memories(
        query="search term",
        limit=10
    )
"""

import json
import time
from datetime import datetime
from typing import Dict, List, Optional, Union, Any, BinaryIO
from urllib.parse import urlencode

import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

__version__ = "1.0.0"
__author__ = "MemorAI Team"
__email__ = "support@memorai.ro"


class MemorAIError(Exception):
    """Base exception for MemorAI SDK errors"""
    
    def __init__(self, message: str, code: str = None, details: Any = None):
        super().__init__(message)
        self.code = code
        self.details = details


class MemorAIAPIError(MemorAIError):
    """Exception for API-related errors"""
    
    def __init__(self, message: str, status_code: int, code: str = None, details: Any = None):
        super().__init__(message, code, details)
        self.status_code = status_code


class MemorAIAuthError(MemorAIAPIError):
    """Exception for authentication errors"""
    pass


class MemorAIRateLimitError(MemorAIAPIError):
    """Exception for rate limit errors"""
    pass


class Memory:
    """Represents a memory object"""
    
    def __init__(self, **kwargs):
        self.id: Optional[str] = kwargs.get('id')
        self.structured_key: Optional[str] = kwargs.get('structuredKey')
        self.content: str = kwargs.get('content', '')
        self.agent_id: str = kwargs.get('agentId', '')
        self.importance: Optional[int] = kwargs.get('importance')
        self.project: Optional[str] = kwargs.get('project')
        self.tags: List[str] = kwargs.get('tags', [])
        self.created_at: Optional[str] = kwargs.get('createdAt')
        self.updated_at: Optional[str] = kwargs.get('updatedAt')
        self.metadata: Dict[str, Any] = kwargs.get('metadata', {})
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert memory to dictionary"""
        return {
            'id': self.id,
            'structuredKey': self.structured_key,
            'content': self.content,
            'agentId': self.agent_id,
            'importance': self.importance,
            'project': self.project,
            'tags': self.tags,
            'createdAt': self.created_at,
            'updatedAt': self.updated_at,
            'metadata': self.metadata
        }
    
    def __repr__(self):
        return f"Memory(id='{self.id}', content='{self.content[:50]}...', agent_id='{self.agent_id}')"


class MemorAI:
    """Official Python client for MemorAI REST API"""
    
    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.memorai.ro",
        timeout: int = 30,
        max_retries: int = 3,
        user_agent: str = None
    ):
        """
        Initialize MemorAI client
        
        Args:
            api_key: Your MemorAI API key
            base_url: Base URL for the API
            timeout: Request timeout in seconds
            max_retries: Maximum number of retries for failed requests
            user_agent: Custom user agent string
        """
        if not api_key:
            raise MemorAIError("API key is required", "MISSING_API_KEY")
        
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.user_agent = user_agent or f"MemorAI-Python-SDK/{__version__}"
        
        # Setup session with retries
        self.session = requests.Session()
        
        retry_strategy = Retry(
            total=max_retries,
            backoff_factor=1,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["HEAD", "GET", "OPTIONS", "POST", "PUT", "DELETE"]
        )
        
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)
        
        # Set default headers
        self.session.headers.update({
            'Authorization': f'Bearer {self.api_key}',
            'Content-Type': 'application/json',
            'User-Agent': self.user_agent
        })
    
    def _request(
        self,
        method: str,
        endpoint: str,
        params: Dict[str, Any] = None,
        data: Any = None,
        files: Dict[str, Any] = None,
        stream: bool = False
    ) -> requests.Response:
        """Make HTTP request to API"""
        url = f"{self.base_url}{endpoint}"
        
        kwargs = {
            'timeout': self.timeout,
            'params': params,
            'stream': stream
        }
        
        if files:
            kwargs['files'] = files
            if data:
                kwargs['data'] = data
        elif data is not None:
            kwargs['json'] = data
        
        try:
            response = self.session.request(method, url, **kwargs)
            
            # Handle authentication errors
            if response.status_code == 401:
                raise MemorAIAuthError(
                    "Authentication failed. Check your API key.",
                    response.status_code,
                    "AUTH_FAILED"
                )
            
            # Handle rate limiting
            if response.status_code == 429:
                raise MemorAIRateLimitError(
                    "Rate limit exceeded. Please retry later.",
                    response.status_code,
                    "RATE_LIMIT_EXCEEDED"
                )
            
            # Handle other HTTP errors
            if not response.ok:
                try:
                    error_data = response.json()
                    error_msg = error_data.get('error', {}).get('message', f'HTTP {response.status_code}')
                    error_code = error_data.get('error', {}).get('code', 'API_ERROR')
                    error_details = error_data.get('error', {}).get('details')
                except (json.JSONDecodeError, KeyError):
                    error_msg = f'HTTP {response.status_code}: {response.reason}'
                    error_code = 'HTTP_ERROR'
                    error_details = None
                
                raise MemorAIAPIError(error_msg, response.status_code, error_code, error_details)
            
            return response
            
        except requests.exceptions.RequestException as e:
            raise MemorAIError(f"Request failed: {str(e)}", "NETWORK_ERROR", {"original_error": str(e)})
    
    def create_memory(
        self,
        content: str,
        agent_id: str,
        importance: Optional[int] = None,
        project: Optional[str] = None,
        tags: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> Memory:
        """
        Create a new memory
        
        Args:
            content: Memory content
            agent_id: Agent identifier
            importance: Importance score (0-10)
            project: Project name
            tags: List of tags
            metadata: Additional metadata
            
        Returns:
            Created Memory object
        """
        data = {
            'content': content,
            'agentId': agent_id,
            'importance': importance,
            'project': project,
            'tags': tags or [],
            'metadata': metadata or {}
        }
        
        response = self._request('POST', '/api/memories', data=data)
        result = response.json()
        
        if not result.get('success'):
            error = result.get('error', {})
            raise MemorAIError(error.get('message', 'Failed to create memory'), error.get('code'))
        
        return Memory(**result['data'])
    
    def get_memories(
        self,
        category: Optional[str] = None,
        tags: Optional[List[str]] = None,
        limit: Optional[int] = None
    ) -> List[Memory]:
        """
        Get all memories with optional filtering
        
        Args:
            category: Filter by category
            tags: Filter by tags
            limit: Maximum number of memories to return
            
        Returns:
            List of Memory objects
        """
        params = {}
        if category:
            params['category'] = category
        if tags:
            params['tags'] = ','.join(tags)
        if limit:
            params['limit'] = str(limit)
        
        response = self._request('GET', '/api/memories', params=params)
        result = response.json()
        
        if not result.get('success'):
            error = result.get('error', {})
            raise MemorAIError(error.get('message', 'Failed to fetch memories'), error.get('code'))
        
        return [Memory(**memory) for memory in result['data']]
    
    def get_memory(self, memory_id: str) -> Memory:
        """
        Get a specific memory by ID
        
        Args:
            memory_id: Memory ID
            
        Returns:
            Memory object
        """
        response = self._request('GET', f'/api/memories/{memory_id}')
        result = response.json()
        
        if not result.get('success'):
            error = result.get('error', {})
            raise MemorAIError(error.get('message', 'Failed to fetch memory'), error.get('code'))
        
        return Memory(**result['data'])
    
    def update_memory(
        self,
        memory_id: str,
        content: Optional[str] = None,
        importance: Optional[int] = None,
        project: Optional[str] = None,
        tags: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> Memory:
        """
        Update a memory
        
        Args:
            memory_id: Memory ID
            content: Updated content
            importance: Updated importance score
            project: Updated project name
            tags: Updated tags
            metadata: Updated metadata
            
        Returns:
            Updated Memory object
        """
        data = {}
        if content is not None:
            data['content'] = content
        if importance is not None:
            data['importance'] = importance
        if project is not None:
            data['project'] = project
        if tags is not None:
            data['tags'] = tags
        if metadata is not None:
            data['metadata'] = metadata
        
        response = self._request('PUT', f'/api/memories/{memory_id}', data=data)
        result = response.json()
        
        if not result.get('success'):
            error = result.get('error', {})
            raise MemorAIError(error.get('message', 'Failed to update memory'), error.get('code'))
        
        return Memory(**result['data'])
    
    def delete_memory(self, memory_id: str) -> bool:
        """
        Delete a memory
        
        Args:
            memory_id: Memory ID
            
        Returns:
            True if deletion was successful
        """
        response = self._request('DELETE', f'/api/memories/{memory_id}')
        result = response.json()
        
        if not result.get('success'):
            error = result.get('error', {})
            raise MemorAIError(error.get('message', 'Failed to delete memory'), error.get('code'))
        
        return result['data'].get('deleted', False)
    
    def search_memories(
        self,
        query: str,
        limit: Optional[int] = None,
        project: Optional[str] = None,
        tags: Optional[List[str]] = None,
        importance_min: Optional[int] = None,
        importance_max: Optional[int] = None,
        date_from: Optional[str] = None,
        date_to: Optional[str] = None
    ) -> List[Memory]:
        """
        Search memories
        
        Args:
            query: Search query
            limit: Maximum number of results
            project: Filter by project
            tags: Filter by tags
            importance_min: Minimum importance score
            importance_max: Maximum importance score
            date_from: Start date (ISO format)
            date_to: End date (ISO format)
            
        Returns:
            List of Memory objects
        """
        params = {'q': query}
        if limit:
            params['limit'] = str(limit)
        if project:
            params['project'] = project
        if tags:
            params['tags'] = ','.join(tags)
        if importance_min is not None:
            params['importanceMin'] = str(importance_min)
        if importance_max is not None:
            params['importanceMax'] = str(importance_max)
        if date_from:
            params['dateFrom'] = date_from
        if date_to:
            params['dateTo'] = date_to
        
        response = self._request('GET', '/api/search', params=params)
        result = response.json()
        
        if not result.get('success'):
            error = result.get('error', {})
            raise MemorAIError(error.get('message', 'Failed to search memories'), error.get('code'))
        
        return [Memory(**memory) for memory in result['data']]
    
    def export_memories(
        self,
        format: str = 'json',
        project: Optional[str] = None,
        tags: Optional[List[str]] = None,
        date_from: Optional[str] = None,
        date_to: Optional[str] = None
    ) -> bytes:
        """
        Export memories
        
        Args:
            format: Export format ('json' or 'csv')
            project: Filter by project
            tags: Filter by tags
            date_from: Start date (ISO format)
            date_to: End date (ISO format)
            
        Returns:
            Export data as bytes
        """
        params = {'format': format}
        if project:
            params['project'] = project
        if tags:
            params['tags'] = ','.join(tags)
        if date_from:
            params['dateFrom'] = date_from
        if date_to:
            params['dateTo'] = date_to
        
        response = self._request('GET', '/api/memories/export', params=params, stream=True)
        return response.content
    
    def import_memories(
        self,
        memories: List[Dict[str, Any]],
        skip_duplicates: bool = True,
        update_existing: bool = False,
        preserve_timestamps: bool = False
    ) -> Dict[str, Any]:
        """
        Import memories
        
        Args:
            memories: List of memory dictionaries
            skip_duplicates: Skip duplicate memories
            update_existing: Update existing memories
            preserve_timestamps: Preserve original timestamps
            
        Returns:
            Import summary dictionary
        """
        data = {
            'memories': memories,
            'options': {
                'skipDuplicates': skip_duplicates,
                'updateExisting': update_existing,
                'preserveTimestamps': preserve_timestamps
            }
        }
        
        response = self._request('POST', '/api/memories/import', data=data)
        result = response.json()
        
        if not result.get('success'):
            error = result.get('error', {})
            raise MemorAIError(error.get('message', 'Failed to import memories'), error.get('code'))
        
        return result['data']
    
    def get_health(self) -> Dict[str, Any]:
        """
        Get API health status
        
        Returns:
            Health status dictionary
        """
        response = self._request('GET', '/api/health')
        result = response.json()
        
        if not result.get('success'):
            error = result.get('error', {})
            raise MemorAIError(error.get('message', 'Failed to get health status'), error.get('code'))
        
        return result['data']
    
    def close(self):
        """Close the session"""
        self.session.close()
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()


# Convenience functions
def create_client(api_key: str, **kwargs) -> MemorAI:
    """Create a MemorAI client instance"""
    return MemorAI(api_key=api_key, **kwargs)


# Export main classes
__all__ = [
    'MemorAI',
    'Memory',
    'MemorAIError',
    'MemorAIAPIError',
    'MemorAIAuthError',
    'MemorAIRateLimitError',
    'create_client'
]
